### PR TITLE
HS-1384: Improve stability/recovery of gRPC connections

### PR DESCRIPTION
**Pull request by j-white**
JIRA: https://opennms.atlassian.net/browse/HS-1384

This doesn't solve the problem completely, but does address some known stability issues which I believe are worth getting in while we continue to dive deeper.

Changes include:
* Upgrade gRPC from 1.43.2 to 1.54.1 (includes a Netty upgrade and addresses some race conditions)
* Address synchronization problems in the SimpleReconnectStrategy class

### CODE REVIEW

File: minion/minion-grpc/grpc-client/src/main/java/org/opennms/horizon/minion/grpc/SimpleReconnectStrategy.java [hunk 1]
```diff
@@ -22,18 +25,24 @@ public SimpleReconnectStrategy(ManagedChannel channel, Runnable onConnect, Runna
     }
 
     @Override
-    public void activate() {
+    public synchronized void activate() {
+        if (reconnectTask != null) {
+            // Silently ignore additional activate requests if one is already in progress
```
Comment (deejgregor) [inline minion/minion-grpc/grpc-client/src/main/java/org/opennms/horizon/minion/grpc/SimpleReconnectStrategy.java:30]: Is there any value in knowing when we reached this block? a debug or a trace message?

(I'm not familiar at all with this to know, but I always worry a little bit when we silently ignore things, as those can be messy to debug when something odd happens)

Comment (j-white) [reply minion/minion-grpc/grpc-client/src/main/java/org/opennms/horizon/minion/grpc/SimpleReconnectStrategy.java:30]: It's normal/expected for this to be called multiple times, so nothing worth logging really.
Could add something on DEBUG/TRACE if we really want to know though.

File: minion/minion-grpc/grpc-client/src/main/java/org/opennms/horizon/minion/grpc/SimpleReconnectStrategy.java [hunk 2]
```diff
@@ -22,18 +25,24 @@ public SimpleReconnectStrategy(ManagedChannel channel, Runnable onConnect, Runna
     }
 
     @Override
-    public void activate() {
+    public synchronized void activate() {
```
Comment (artnaseef) [inline minion/minion-grpc/grpc-client/src/main/java/org/opennms/horizon/minion/grpc/SimpleReconnectStrategy.java:28]: These 2 methods using `synchronized` keyword with callbacks are creating coarse critical sections (i.e. not light and fast).  The `onConnect` callback makes a GRPC method call to create a stream, and sends a message over that stream.

Comment (j-white) [reply minion/minion-grpc/grpc-client/src/main/java/org/opennms/horizon/minion/grpc/SimpleReconnectStrategy.java:28]: Good point. Will need a different approach then.

**Review by artnaseef — APPROVED** [review]

**Pull request merged**